Docs/KDESyntaxDefinition/dafny.xml

```diff
@@ -0,0 +1,151 @@
+      <item>static</item>
+      <item>then</item>
+      <item>this</item>
+      <item>trait</item>
+      <item>true</item>
+      <item>twostate</item>
+      <item>type</item>
+      <item>unchanged</item>
+      <item>var</item>
+      <item>where</item>
+      <item>while</item>
+      <item>yield</item>
+      <item>yields</item>
+    </list>
+    <list name="types">
+      <item>array</item>
+      <item>bool</item>
+      <item>char</item>
+      <item>imap</item>
+      <item>int</item>
+      <item>iset</item>
+      <item>map</item>
+      <item>multiset</item>
+      <item>nat</item>
+      <item>object</item>
+      <item>real</item>
+      <item>seq</item>
+      <item>set</item>
+      <item>string</item>
+    </list>
+
+    <list name="verification keywords">
+      <item>decreases</item>
+      <item>ensures</item>
+      <item>invariant</item>
+      <item>modifies</item>
+      <item>provides</item>
+      <item>reads</item>
+      <item>requires</item>
+      <item>reveals</item>
+    </list>
+
+    <contexts>
+      <context name="Normal" attribute="Normal Text" lineEndContext="#stay">
+        <keyword attribute="Keyword" String="keywords" context="#stay" />
+        <keyword attribute="Types" String="types" context="#stay" />
+        <keyword attribute="Assertions" String="verification keywords" context="#stay" />
+
+        <RegExpr attribute="Types" String="bv0|bv[1-9]\d*"/>
+        <RegExpr attribute="Types" String="array[2-9]\d*"/>
+
+        <Int attribute="Decimal" context="#stay" />
+        <Float attribute="Float" context="#stay" />
+        <HlCChar attribute="Char" context="#stay" />
+        <DetectChar attribute="String" char="&quot;" context="Quoted String" />
+
+        <Detect2Chars attribute="Comment" char="/" char1="/" context="SingleLineComment" />
+        <Detect2Chars attribute="Comment" char="/" char1="*" context="MultiLineComment" />
+      </context>
+
+      <context name="Quoted String" attribute="String" lineEndContext="#pop">
+        <Detect2Chars attribute="String" context="#stay" char="\\" char1="&quot;"/>
+        <DetectChar attribute="String" char="&quot;" context="#pop" />
+      </context>
+
+      <context name="SingleLineComment" attribute="Comment" lineEndContext="#pop"/>
+      <context name="MultiLineComment" attribute="Comment" lineEndContext="#stay">
+        <Detect2Chars attribute="Comment" context="#pop" char="*" char1="/"/>
+      </context>
+      <context name="Documentation" attribute="Comment" lineEndContext="#pop" />
+    </contexts>
+
+    <itemDatas>
+      <itemData name="Normal Text" defStyleNum="dsNormal" />
+
+      <itemData name="Keyword" defStyleNum="dsKeyword" />
+      <itemData name="Types" defStyleNum="dsDataType" />
+      <itemData name="Assertions" defStyleNum="dsOthers" />
+
+      <itemData name="Decimal" defStyleNum="dsDecVal" />
+      <itemData name="Float" defStyleNum="dsFloat" />
+      <itemData name="Char" defStyleNum="dsChar" />
```

Comment:
I don't understand what this does, but I'm curious if this handles Dafny trickery around lexing `char` literals? (I'm not going to insist that the syntax highlighter can do everything that a flexible scanner/parser can do, but it would be nice if it does.)

The tricky thing is that `'` (single quote) can be part of identifiers. In fact, an identifier can even begin and end with a `'`. Essentially, the rule is that if something can be a character literal, then it is; otherwise, it is an identifier. Examples as identifiers:

    'a
    a'
    'ab'
    'a'b'
    'a''b'

Examples of character literals:

    'a'
    '\n'
    '\0'
    '\u1234'
    '\uabcd'
    '\uAbC8'
    '"'
    '\''

Example of something illegal:

    '\u123'

Strings also have a complicated syntax. What goes between the `"` and `"` is a sequence of the things that can go between the `'` and `'` in a `char` literal (there are exceptions around `"` and `'`, but I won't give the details here). Also, Dafny has _verbatim strings_, like in C#.

Reply:
It was tricky to do, but I got it to handle Dafny char literals.